x-pack/plugins/reporting/server/lib/task.ts

```diff
@@ -0,0 +1,290 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import moment from 'moment';
+import { ReportingCore } from '../';
+import {
+  ConcreteTaskInstance,
+  RunContext,
+  TaskManagerSetupContract,
+  TaskManagerStartContract,
+} from '../../../task_manager/server';
+import { CancellationToken } from '../../common';
+import { ESQueueWorkerExecuteFn, TaskRunResult } from '../types';
+import { LevelLogger, statuses } from './';
+import { ReportingStore, ReportTaskJSON } from './store';
+
+const REPORTING_QUEUE_JOB_TYPE = 'reporting-queue-job';
+
+export interface TaskParams<JobParamsType> {
+  report: ReportTaskJSON;
+  payload: JobParamsType;
+  user: string | null; // user is optional as security may be disabled
+  headers: any;
+}
+
+export type SchedulingFn = <JobParamsType>(
+  params: TaskParams<JobParamsType>
+) => Promise<ConcreteTaskInstance>;
+
+interface ReportingTaskInstanceParams<JobParamsType> {
+  taskType: string;
+  state: object;
+  params: TaskParams<JobParamsType>;
+}
+
+/*
```

Comment:
NOTE: this file is a re-write of `x-pack/plugins/reporting/server/lib/esqueue/worker.js`